tensorflow/python/keras/layers/normalization.py

```diff
@@ -1010,16 +1034,60 @@ def _broadcast(v):
           self.axis != [ndims - 1]):
         return array_ops.reshape(v, broadcast_shape)
       return v
-    scale, offset = _broadcast(self.gamma), _broadcast(self.beta)
 
-    # Compute layer normalization using the batch_normalization function.
-    outputs = nn.batch_normalization(
-        inputs,
-        mean,
-        variance,
-        offset=offset,
-        scale=scale,
-        variance_epsilon=self.epsilon)
+    if not self.fused:
+      # Calculate the moments on the last axis (layer activations).
+      mean, variance = nn.moments(inputs, self.axis, keep_dims=True)
+
+      scale, offset = _broadcast(self.gamma), _broadcast(self.beta)
+
+      # Compute layer normalization using the batch_normalization function.
+      outputs = nn.batch_normalization(
+          inputs,
+          mean,
+          variance,
+          offset=offset,
+          scale=scale,
+          variance_epsilon=self.epsilon)
+    else:
+      # Collapse dims before self.axis, and dims in self.axis
+      pre_dim, in_dim = (None, None)
+      axis = sorted(self.axis)
+      tensor_shape = array_ops.shape(inputs)
+      for dim in range(0, ndims):
+        dim_tensor = tensor_shape[dim]
+        if dim < axis[0]:
+          pre_dim = dim_tensor if pre_dim is None else (pre_dim * dim_tensor)
+        elif dim in axis:
+          in_dim = dim_tensor if in_dim is None else (in_dim * dim_tensor)
+      
+      squeezed_shape = [1, pre_dim, in_dim, 1]
+      data_format = 'NCHW'
+
+      inputs = array_ops.reshape(inputs, squeezed_shape)
+
+      def _set_const_tensor(val, dtype, shape):
+        return array_ops.fill(shape, constant_op.constant(val, dtype=dtype))
+
+      scale = _set_const_tensor(1.0, inputs.dtype, [pre_dim])
+      offset = _set_const_tensor(0.0, inputs.dtype, [pre_dim])
+
+      # Compute layer normalization using the fused_batch_norm function.
+      outputs, _, _ = nn.fused_batch_norm(
+          inputs,
+          scale=scale,
```

Comment:
I don't think we can reshape gamma and beta as input to the fused_batch_norm. Basically, batch norm requires mean/var has the same shape with gamma/beta. Suppose the input the NxD and the batch norm is over N, both mean/var and gamma/beta are 1xD. However, by definition, the layer norm is over D and the mean/var is Nx1 but gamma/beta is 1xD. To utilize the fused batch norm in the layer norm, we let the fused batch norm to normalize over D (instead of N), which will require mean/var and gamma/beta in shape of Nx1. However, our gamma/beta is 1xD and we cannot reshape it to Nx1. So, our strategy is to give a Nx1 const gamma/beta (1, 0) to do a "redundant" scale and offset. And then separately, we handle the meaningful scale and offset using the real 1xD gamma/beta.

Reply:
Can you add a comment? Something like "self.gamma and self.beta have the wrong shape for fused_batch_norm, so we cannot pass them as the scale and offset parameters."